subsys/pm/policy.c

```diff
@@ -0,0 +1,165 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pm/pm.h>
+#include <pm/policy.h>
+#include <spinlock.h>
+#include <sys_clock.h>
+#include <sys/__assert.h>
+#include <sys/time_units.h>
+#include <sys/atomic.h>
+
+/** State disallow reference counting */
+static atomic_t state_disallow_cnt[PM_STATE_COUNT];
+/** Callback to notify when state allowed status changes */
+static pm_policy_state_allowed_changed_cb_t allowed_changed_cb;
+
+/** Lock to synchronize access to the latency request list. */
+static struct k_spinlock latency_lock;
+/** List of maximum latency requests. */
+static sys_slist_t latency_reqs;
+/** Maximum CPU latency in ticks */
+static int32_t max_latency_ticks = K_TICKS_FOREVER;
+/** Callback to notify when maximum latency changes. */
+pm_policy_latency_changed_cb_t latency_changed_cb;
+
+/** @brief Update maximum allowed latency. */
+static void update_max_latency(void)
+{
+	int32_t new_max_latency_ticks = K_TICKS_FOREVER;
+	struct pm_policy_latency_request *req;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&latency_reqs, req, node) {
+		if ((new_max_latency_ticks == K_TICKS_FOREVER) ||
+		    ((int32_t)req->value < new_max_latency_ticks)) {
+			new_max_latency_ticks = (int32_t)req->value;
+		}
+	}
+
+	if ((latency_changed_cb != NULL) &&
+	    (max_latency_ticks != new_max_latency_ticks)) {
+		int32_t latency_us;
+
+		if (new_max_latency_ticks == K_TICKS_FOREVER) {
+			latency_us = SYS_FOREVER_US;
+		} else {
+			latency_us = (int32_t)k_ticks_to_us_ceil32(new_max_latency_ticks);
+		}
+
+		latency_changed_cb(latency_us);
+	}
+
+	max_latency_ticks = new_max_latency_ticks;
+}
+
+#ifdef CONFIG_PM_POLICY_DEFAULT
+const struct pm_state_info *pm_policy_next_state(uint8_t cpu, int32_t ticks)
+{
+	uint8_t num_cpu_states;
+	const struct pm_state_info *cpu_states;
+
+	num_cpu_states = pm_state_cpu_get_all(cpu, &cpu_states);
+
+	for (int16_t i = (int16_t)num_cpu_states - 1; i >= 0; i--) {
+		const struct pm_state_info *state = &cpu_states[i];
+		uint32_t min_residency, exit_latency, total_latency;
+
+		if (!pm_policy_state_is_allowed(state->state)) {
+			continue;
+		}
+
+		min_residency = k_us_to_ticks_ceil32(state->min_residency_us);
+		exit_latency = k_us_to_ticks_ceil32(state->exit_latency_us);
+		total_latency = min_residency + exit_latency;
+
+		/* skip state if it brings too much latency */
+		if ((max_latency_ticks != K_TICKS_FOREVER) &&
+		    (total_latency >= max_latency_ticks)) {
+			continue;
+		}
+
+		if ((ticks == K_TICKS_FOREVER) || (ticks >= total_latency)) {
+			return state;
+		}
+	}
+
+	return NULL;
+}
+#endif
+
+void pm_policy_state_disallow(enum pm_state state)
+{
+	if ((atomic_inc(&state_disallow_cnt[state]) == 0) &&
+	    (allowed_changed_cb != NULL)) {
+		allowed_changed_cb(state, false);
+	}
+}
+
+void pm_policy_state_allow(enum pm_state state)
+{
+	atomic_t cnt = atomic_dec(&state_disallow_cnt[state]);
+
+	__ASSERT(cnt >= 1, "Unbalanced state disallow/allow");
```

Comment:
but why is it not allowed? what if two modules disallow at the same time and later one of them lifts the restriction?

if the assert is left only one module would be able to keep the restriction at any given time which seems like a huge limitation.

Reply:
Calling `allow` without a previous `disallow` is wrong, so an unbalanced sequence should never happen in well-behaved code.

Reply:
Ah... yes. I read it the wrong way around. For some reason though this is restricting multiple disallow.
It is ok obviously :) Sorry